confmap/expand.go

```diff
@@ -0,0 +1,206 @@
+
+	errURILimit = errors.New("reached limit of 100 URIs")
+)
+
+func (mr *Resolver) expandValueRecursively(ctx context.Context, value any) (any, error) {
+	for i := 0; i < 100; i++ {
+		val, changed, err := mr.expandValue(ctx, value)
+		if err != nil {
+			return nil, err
+		}
+		if !changed {
+			return val, nil
+		}
+		value = val
+	}
+	return nil, errTooManyRecursiveExpansions
+}
+
+func (mr *Resolver) expandValue(ctx context.Context, value any) (any, bool, error) {
+	switch v := value.(type) {
+	case string:
+		if !strings.Contains(v, "${") || !strings.Contains(v, "}") {
+			// No URIs to expand.
+			return value, false, nil
+		}
+
+		if strings.Count(v, "}") > 100 {
+			// Too many closing brackets. Don't expand to protect from too deep recursion.
+			return "", false, errURILimit
+		}
+
+		uri := findURI(v)
+		if uri != "" && uri == value {
+			// If the value is a single URI, then the return value can be anything.
+			// This is the case `foo: ${file:some_extra_config.yml}`.
+			return mr.expandURI(ctx, v)
+		}
+
+		// Embedded or nested URIs.
+		return mr.findAndExpandURI(ctx, v)
+	case []any:
+		nslice := make([]any, 0, len(v))
+		nchanged := false
+		for _, vint := range v {
+			val, changed, err := mr.expandValue(ctx, vint)
+			if err != nil {
+				return nil, false, err
+			}
+			nslice = append(nslice, val)
+			nchanged = nchanged || changed
+		}
+		return nslice, nchanged, nil
+	case map[string]any:
+		nmap := map[string]any{}
+		nchanged := false
+		for mk, mv := range v {
+			val, changed, err := mr.expandValue(ctx, mv)
+			if err != nil {
+				return nil, false, err
+			}
+			nmap[mk] = val
+			nchanged = nchanged || changed
+		}
+		return nmap, nchanged, nil
+	}
+	return value, false, nil
+}
+
+// findURI finds the URI corresponding to the first closing bracket in input. It returns
+// the URI if it is expandable, or an empty string if it is not expandable.
+// Note: findURI is only called when input contains a closing bracket.
+func findURI(input string) string {
+	closeIndex := strings.Index(input, "}")
+	openIndex := strings.LastIndex(input[:closeIndex+1], "${")
+	if openIndex < 0 {
+		// Should not expand because there is a missing ${.
+		return ""
+	}
+
+	uri := input[openIndex : closeIndex+1]
+
+	if !strings.Contains(uri, ":") {
+		// Should not expand. This is expanded in the expandconverter.
+		return ""
+	}
+
+	return uri
+}
+
+// findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input.
+func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (output string, changed bool, err error) {
+	var repl string
+	var expandedRemaining string
+
+	uri := findURI(input)
+	if uri == "" {
+		// The first URI in input is not expandable. Strip the first URI from input and check if
+		// other URIs are expandable.
+		closeIndex := strings.Index(input, "}")
+		noExpand := input[:closeIndex+1]
+		remaining := input[closeIndex+1:]
```

Comment:
This code is duplicate with findURI. Let's do your suggestion to change findURI to be recursive.